File: flow/texture.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/flow/texture.h"

namespace flow {

TextureRegistry::TextureRegistry() = default;

TextureRegistry::~TextureRegistry() = default;

size_t TextureRegistry::RegisterTexture(std::shared_ptr<Texture> texture) {
  ASSERT_IS_GPU_THREAD
  size_t id = counter_++;
  mapping_[id] = texture;
  texture->id_ = id;
  return id;
}

void TextureRegistry::UnregisterTexture(size_t id) {
  ASSERT_IS_GPU_THREAD
  mapping_.erase(id);
}

void TextureRegistry::OnGrContextCreated() {
  ASSERT_IS_GPU_THREAD;
  for (auto& it : mapping_) {
    it.second->OnGrContextCreated();
  }
}

void TextureRegistry::OnGrContextDestroyed() {
  ASSERT_IS_GPU_THREAD;
  for (auto& it : mapping_) {
    it.second->OnGrContextDestroyed();
  }
}

std::shared_ptr<Texture> TextureRegistry::GetTexture(size_t id) {
  ASSERT_IS_GPU_THREAD
  return mapping_[id];
}

Texture::~Texture() = default;

}  // namespace flow

File: lib/web_ui/lib/src/engine/canvaskit/texture.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:ui/ui.dart' as ui;

import '../texture.dart';
import 'canvaskit_api.dart';
import 'image.dart';
import 'layer.dart';
import 'painting.dart';

class CkTexture {
  CkTexture(this.source);

  final Object source;

  CkImage? _ckImage;

  bool _newFrameReady = false;

  final CkPaint _paint = CkPaint();

  void markNewFrameAvailable() {
    _newFrameReady = true;
  }

  void paint(
    PaintContext context,
    ui.Offset offset,
    double width,
    double height,
    bool freeze,
    ui.FilterQuality filterQuality,
  ) {
    if (_ckImage == null || _newFrameReady && !freeze) {
      final SkImage? skImage =
          canvasKit.MakeLazyImageFromTextureSource(source, null);

      if (skImage == null) {
        return;
      }

      _ckImage?.dispose();
      _ckImage = CkImage(skImage);
    }

    context.leafNodesCanvas?.drawImageRect(
      _ckImage!,
      ui.Rect.fromLTWH(
        0,
        0,
        _ckImage!.width.toDouble(),
        _ckImage!.height.toDouble(),
      ),
      ui.Rect.fromLTWH(offset.dx, offset.dy, width, height),
      _paint..filterQuality = filterQuality,
    );
  }

  void dispose() {
    _ckImage?.dispose();
  }
}

class CkTextureRegistry implements TextureRegistry {
  CkTextureRegistry._();

  static final CkTextureRegistry instance = CkTextureRegistry._();

  final Map<int, CkTexture> _textures = <int, CkTexture>{};

  int _nextTextureId = 0;

  @override
  int registerTexture(Object source) {
    final int id = _nextTextureId++;
    _textures[id] = CkTexture(source);
    return id;
  }

  @override
  void unregisterTexture(int id) {
    final CkTexture? texture = _textures.remove(id);
    texture?.dispose();
  }

  CkTexture? getTexture(int id) {
    return _textures[id];
  }

  @override
  void textureFrameAvailable(int id) {
    final CkTexture? texture = _textures[id];
    if (texture != null) {
      texture.markNewFrameAvailable();
    }
  }
}
